docs/dev/system_keyspace.md

```diff
@@ -149,6 +149,51 @@ merge the data in the legacy store with data the `truncated` table. Until the wh
 cluster agrees on the feature `TRUNCATION_TABLE` truncation will write both new and 
 legacy records. When the feature is agreed upon the legacy map is removed.
 
+## system.quarantined_hosts
+
+Holds a list of host identifiers of nodes that were removed from the cluster and are not allowed to join back.
+
+Node operations like remove node, replace node, or decommissioned reassign token ownership of the target node.
+After such operations, the removed host must not rejoin the cluster as it might hold stale data, causing data resurrection.
+
+Schema:
+~~~
+CREATE TABLE system.quarantined_hosts (
+    key text,
+    host_id uuid,
+    quarantined_at timestamp,
+    PRIMARY KEY (key, host_id)
+) WITH CLUSTERING ORDER BY (host_id ASC)
+~~~
+
+The table consists of a single partition (where `key = 'QUARANTINED_HOSTS'`)
+and each quarantined host is stored in a row, with the `host_id` as the clustering key.
+The time when the host was quarantined is stored in `quarantined_at`.
+
+When node operations like `removenode`, `replace`, or `decommission` complete,
+the target hosts are quarantined from the token ring.
+Their `host_id` is inserted to `system.quarantined_hosts` and the `QUARANTINED_HOSTS` application state is updated and gossiped to all other nodes
+(as done on startup when gossip loads `system.quarantined_hosts` to initialize said application state).
+
+Note that the quarantine is restricted at this stage only to attempts to re-join the ring after a restart.
+If the said nodes are running all this time behind a network partition
+and the network partition heals after the nodes were removed or replaced (using a different ip address),
+quarantining using this mechanism will have no effect.
```

Comment:
See https://github.com/scylladb/scylladb/pull/12040/files, we could pass the host_id (unique version) as part of the gossip message and block any gossip messages with the blocked host_id.

This makes network partition case work.

Reply:
For gossip, but not for RPCs and data queries in particular.
For example, the node could be retrying truncate, or write after the data was deleted and the tombstone purged (with low gc_grace_seconds), causing resurrection.

Reply:
This is an interesting idea. However, if I understand correctly, this could be done in a follow-up. In that case I suggest we don't extend the scope of the PR further.